nextcord/http.py

```diff
@@ -155,25 +155,286 @@ def bucket(self) -> str:
         return f"{self.channel_id}:{self.guild_id}:{self.path}"
 
 
-class MaybeUnlock:
-    def __init__(self, lock: asyncio.Lock) -> None:
-        self.lock: asyncio.Lock = lock
-        self._unlock: bool = True
+class RateLimitMigrating(DiscordException):
+    ...
+
+
+class IncorrectBucket(DiscordException):
+    ...
+
+
+class RateLimit:
+    def __init__(self, time_offset: float = 0.3) -> None:
+        """Used to time gate a large batch of requests to only occur X every Y seconds. Used via ``async with``
+
+        NOT THREAD SAFE.
+
+        Parameters
+        ----------
+        time_offset: :class:`float`
+            Number in seconds to increase all timers by. Used for lag compensation.
```

Comment:
This seems to default to quite an arbitrary value and be a bit unnecessary. `unsync_clock` meant that `datetime.now()` was not trustable (out of sync, so it would send requests too early or late), and so `X-RateLimit-Reset` was also not able to be used.

`x_reset_after = float(x_reset_after) + self._time_offset`
If there was a weird latency difference one way to the other, `reset_after` would be adjusted to be longer, but surely it has already decreased by the time the header was received.

If Discord sends `4.5` as a reset-after, and it takes `0.3`s to get back to the client, the reset after is actually `4.2` now since `0.3`s has already elapsed. So it is safe to not add an "offset".

Reply:
When initially tested this implementation, before I decided to move it into NC, I kept running into 429's seemingly caused by Discord servers not being properly time synced with one another, but perhaps I need to retest that.

Reply:
After setting it to 0.0 and running a quick test, it seems fine. Perhaps I'm misremembering what happened in the past. I'll default it to 0.0 and keep it in for now unless something happens in testing or you want it removed entirely.